DEPR: Changing default of str.extract(expand=False) to str.extract(expand=True)

datapythonista opened the pull request:
…o str.extract(expand=True) (#6581)

- [X] related to #6581
- [X] tests added / passed
- [X] passes `git diff upstream/master -u -- "*.py" | flake8 --diff`
- [X] whatsnew entry

jreback commented on doc/source/whatsnew/v0.23.0.txt:
```diff
@@ -314,6 +314,7 @@ Removal of prior version deprecations/changes
 - The ``Panel4D`` and ``PanelND`` classes have been removed (:issue:`13776`)
 - The ``Panel``class has dropped the ``to_long``and ``toLong`` methods (:issue:`19077`)
 - The options ``display.line_with`` and ``display.height`` are removed in favor of ``display.width`` and ``display.max_rows`` respectively (:issue:`4391`, :issue:`19107`)
+- The ``expand`` parameter of :func:`str.extract` method is ``True`` by default (it was `False`)
```
has changed, previously was default to ``None``. add the issue number and this PR number. Move this instead to a sub-section (of API breaking changes) where you show an example of this.

datapythonista commented:
I changed the section in the wantsnew, and updated the text and the issue number, but I left the default parameter.

In the signature, `expand` had a default value of `None`, but in practice [1] it was False, as it was also said in the docstring [2]. I'll change the message if you disagree, but I think it makes more sense to say that by default was `False` before.

1. https://github.com/pandas-dev/pandas/blob/master/pandas/core/strings.py#L687
2. https://github.com/pandas-dev/pandas/blob/master/pandas/core/strings.py#L613

datapythonista commented:
Thank you for the comments @jreback I understood it now. If you want to take a look, I think it should be much better, and with the examples now. Let me know if any other change is needed.

TomAugspurger commented on doc/source/whatsnew/v0.23.0.txt:
```diff
@@ -232,6 +232,81 @@ Build Changes
+
+By default, extracting matching patterns from strings with :func:`str.extract` used to return a
+``Series`` if a single group was being extracted (a ``DataFrame`` if more than one group was
+extracted``). Since Pandas 0.23.0 :func:`str.extract` always returns a ``DataFrame``, unless
```
"Since" -> "As of"

TomAugspurger commented on doc/source/whatsnew/v0.23.0.txt:
```diff
@@ -232,6 +232,81 @@ Build Changes
+``expand`` is set to ``False`` (:issue:`11386`).
+
+Also, ``None`` was an accepted value for the ``expand`` parameter (which was equivalent to
+``False``), but since 0.23..0 it raises a ``ValueError``.
```
" but now raises a ``ValueError``.

TomAugspurger reviewed (commented):
Could you update the prose docs at http://pandas-docs.github.io/pandas-docs-travis/text.html#extract-first-match-in-each-subject-extract

datapythonista commented:
Thanks for the comments @TomAugspurger. I made the changes. In text.rst, there was no example with expand unspecified or None, so the text was already correct. I added a quick note about the default value in the warning.

Please let me know if that looks good, or you've got a better idea.

datapythonista commented:
All the comments from the reviews were addressed, right? Just double checking that the PR doesn't need anything from my side.

jorisvandenbossche commented on doc/source/whatsnew/v0.23.0.txt:
```diff
@@ -232,6 +232,81 @@ Build Changes
+    Out [4]:
+    pandas.core.series.Series
+
+    In [5]: s.str.extract('.*(\d\d).*', expand=None)
```
I don't think it is needed to show this one (it just draws attention away from the more important example IMO)

jorisvandenbossche commented on doc/source/whatsnew/v0.23.0.txt:
```diff
@@ -232,6 +232,81 @@ Build Changes
+
+New Behavior:
+
+.. code-block:: ipython
```
Can you make this a `.. ipython:: python` directive? (and remove the output below) ?

datapythonista replied:
Should I use the `..ipython:: python` directive in all the 3 cases, right?

I'm not sure what do you mean by "remove the output below", sorry. Can you clarify please?

jorisvandenbossche replied:
To clarify: you can use it for code that can run *now*, so for the new behaviour. For showing the "old behaviour" you still need to use the literal code-block (there you don't need to change anything). 

And by removing output I mean that in an `ipython:: python` block, you only include the actual code, as the output of the code is generated during the doc build. If you scroll a bit up in the file, you see some examples of that.

jorisvandenbossche reviewed (commented):
Just added some minor comments on the whatsnew notice

datapythonista committed:
DEPR: Changing default of str.extract(expand=False) to str.extract(expand=True) (#6581)

datapythonista commented:
Thanks for the info @jorisvandenbossche, didn't know about the `.. ipython::` directive.

It should be all right now.

jorisvandenbossche committed:
remove whitespace

jorisvandenbossche commented:
@datapythonista Thanks!
